### PR TITLE
mmanon: Simplified and fixed IPv4 digit detection.

### PR DESCRIPTION
**Pull request by alorbach**
- Fixed an issue with numbers above int64 in syntax_ipv4.
  Numbers that were up to 256 above the max of an int64
  could incorrectly be detected as valid ipv4 digit.
- Simplified the IPv4 digit detection function and renamed
  to isPosByte.
- added testcasse for malformed IPvc4 addresses

closes: https://github.com/rsyslog/rsyslog/issues/4940

### CODE REVIEW

File: tests/mmanon_simple_mallformed_ipv4.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+generate_conf
+add_conf '
+template(name="outfmt" type="string" string="%msg%\n")
+
+module(load="../plugins/mmanon/.libs/mmanon")
+module(load="../plugins/imtcp/.libs/imtcp")
+input(type="imtcp" port="0" listenPortFileName="'$RSYSLOG_DYNNAME'.tcpflood_port" ruleset="testing")
+
+ruleset(name="testing") {
+	action(type="mmanon" ipv4.bits="32" ipv4.mode="simple")
```
Comment (micah) [inline tests/mmanon_simple_mallformed_ipv4.sh:20]: I dont think `mode=simple` tests the relevant code path, it should be `mode=zero`, if you look at https://github.com/rsyslog/rsyslog/blob/master/plugins/mmanon/mmanon.c#L785

Comment (alorbach) [reply tests/mmanon_simple_mallformed_ipv4.sh:20]: Indeed, but my main goal for this test was to make confirm that those string did not get detected as IPv4 addresses. The new syntax_ipv4 & isPosByte function implementation should makes sure valid IPv4 addresses will be valid in any case.

Comment (micah) [reply tests/mmanon_simple_mallformed_ipv4.sh:20]: I do agree with you, however... it might be wise to add a similar test with mode=zero, especially since it was not crashing because of that fix.

File: plugins/mmanon/mmanon.c [hunk 2]
```diff
@@ -388,73 +389,83 @@ getHexVal(char c)
...
 {
-	int64_t val;
-	size_t nproc;
+	size_t nproc = 0;
 	size_t i;
 	int r = 0;
-
-	val = getPosInt(buf, buflen, &i);
-	if(val < 0 || val > 255)
+	if(	isPosByte(buf, buflen, &i) == 0) {
```
Comment (micah) [inline plugins/mmanon/mmanon.c]: Just code-quality note here, you added a space after the parens in all the changes you made, but the original code doesn't seem to have those (nor is the space balanced on the other end of the closing parens).

Note: I only point this out here, but it is the same for the other bits you added.

File: plugins/mmanon/mmanon.c [hunk 3]
```diff
@@ -388,73 +389,83 @@ getHexVal(char c)
...
+	}
 	i += nproc;
 
 	*nprocessed = i;
 	r = 1;
-
 done:
+/*
+	DBGPRINTF("syntax_ipv4: %s Is IPV4=%s\n", buf, (r == 1 ? "YES" : "NO"));
+*/
```
Comment (micah) [inline plugins/mmanon/mmanon.c]: Did you mean to leave this code here commented out?

**Review by micah — CHANGES_REQUESTED** [review]

File: plugins/mmanon/mmanon.c (hunk 3 above)
Comment (alorbach) [reply plugins/mmanon/mmanon.c]: > 

yes it is quiet a very busy output which was helpful for debugging only. 
I have removed it along with the extra spaces
